File: packages/core/src/research_engine/adapters/inference/routing.py
"""Where inference runs, and what happens when that place is unreachable.

A single ``RE_EMBEDDING_BASE_URL`` used to decide three unrelated things at
once: where compute happens (a placement decision, and a genuinely dynamic one
— a desktop sleeps), which model is authoritative (a corpus invariant that must
never drift), and whether a failure is fatal (a per-call-site policy). Setting
the URL silently answered all three, and the only way to answer the first one
differently was to delete the address.

This module separates them.

**Placement** is `*_provider`, and the mode wins over the URL's presence, so
`local_bge` is an off switch you can flip without losing the host address.

**Model identity** is not defended here at all, because it is already defended
better elsewhere: `passage_embeddings` is keyed by `(passage_id, model,
model_version)` and search only ever reads rows matching the live embedder. A
wrong-model vector does not corrupt the index, it lands in a namespace nothing
queries. What genuinely needs catching is a backend that *lies* about its
identity, and that is the `/health` handshake's job.

**Failure policy** is the part worth thinking about, and it is not the
read/write split it first appears to be. Falling back to a backend that passes
the same identity handshake is safe — local and remote bge-m3 were measured
bit-identical, `max|Δ| = 0.00e+00`. What differs is the *cost of degrading*:

- A query embedding is one short string. Falling back costs a model load and
  then 66 ms. Search stays up; nobody needs to know.
- A corpus-wide embed is 255k passages. Falling back turns hours on a 3090 into
  days on a laptop. That must stop and say so, not quietly downgrade.
- A rerank has a third option the others do not: skip it. On a host without a
  working accelerator, reranking 30 candidates measured 48.8 s of a 49.1 s
  search. Both obvious answers are bad — failing the search outright, or
  silently spending 49 seconds — so the search degrades in *quality, visibly*
  instead of in *latency, invisibly*.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import TYPE_CHECKING, Any

import structlog

from research_engine.domain.errors import ConfigurationError, EmbeddingUnavailable

if TYPE_CHECKING:
    from collections.abc import Callable, Sequence

    from research_engine.config.settings import Settings
    from research_engine.ports.embedding import EmbeddingPort
    from research_engine.ports.reranker import RerankerPort

logger = structlog.get_logger()


class Workload(StrEnum):
    """What a caller is doing, which is what decides the failure policy."""

    #: One short text, a person waiting. Degrading beats failing.
    QUERY = "query"
    #: Corpus-wide, unattended, results are stored. Failing beats degrading,
    #: because a silent downgrade here costs days and nobody is watching.
    BULK = "bulk"


class QueryEmbeddingWithFallback:
    """Remote embedding that falls back to local rather than failing a search.

    Only ever wired into the query path. The local model is built on first need,
    not up front, so a healthy remote setup never pays the 2.3 GB load.

    Safe because both ends run the model named by `RE_EMBEDDING_MODEL` and the
    remote end has proved it on `/health`; the vectors are interchangeable. It
    is `EmbeddingUnavailable` alone that triggers this — a `ModelMismatch` is a
    misconfiguration, and quietly papering over it is how a corpus ends up half
    embedded by something nobody chose.
    """

    def __init__(
        self,
        remote: EmbeddingPort,
        build_local: Callable[[], EmbeddingPort],
        *,
        base_url: str,
    ) -> None:
        self._remote = remote
        self._build_local = build_local
        self._local: EmbeddingPort | None = None
        self._base_url = base_url
        self._warned = False

    @property
    def model_name(self) -> str:
        return self._remote.model_name

    @property
    def model_version(self) -> str:
        return self._remote.model_version

    @property
    def dim(self) -> int:
        return self._remote.dim

    async def embed(self, text: str) -> list[float]:
        return (await self.embed_batch([text]))[0]

    async def embed_batch(self, texts: Sequence[str]) -> list[list[float]]:
        if not texts:
            return []
        try:
            return await self._remote.embed_batch(list(texts))
        except EmbeddingUnavailable as exc:
            if not self._warned:
                # Once per process. A warning on every keystroke-fast query
                # would bury the one line that explains the slowdown.
                logger.warning(
                    "query_embedding_fell_back_to_local",
                    base_url=self._base_url,
                    error=str(exc),
                    detail="Queries are being embedded on this machine. Corpus "
                    "ingestion still requires the GPU host and will refuse to "
                    "run without it.",
                )
                self._warned = True
            if self._local is None:
                self._local = self._build_local()
            return await self._local.embed_batch(list(texts))


@dataclass
class InferenceBackends:
    """The adapters the composition root wires in, plus how to shut them down."""

    query_embedding: EmbeddingPort
    bulk_embedding: EmbeddingPort
    reranker: RerankerPort
    summary: str
    closeables: list[Any] = field(default_factory=list)

    async def close(self) -> None:
        for target in self.closeables:
            closer = getattr(target, "close", None)
            if closer is None:
                continue
            try:
                await closer()
            except Exception as exc:  # noqa: BLE001 - shutdown must not raise
                logger.warning("inference_close_failed", error=str(exc))


def build_inference(settings: Settings) -> InferenceBackends:
    """Build embedding and reranking adapters according to the configured modes."""
    from research_engine.adapters.embedding.local_bge import LocalBGEEmbedding
    from research_engine.adapters.embedding.remote_api import RemoteEmbeddingClient
    from research_engine.adapters.reranker.local_bge import LocalBGEReranker
    from research_engine.adapters.reranker.noop import NoopReranker
    from research_engine.adapters.reranker.remote_api import RemoteReranker

    base_url = settings.resolved_inference_base_url
    if settings.inference_base_url is None and settings.embedding_base_url:
        logger.info(
            "inference_base_url_deprecated_alias",
            detail="RE_EMBEDDING_BASE_URL still works; RE_INFERENCE_BASE_URL is "
            "the current name, since one server now serves both models.",
        )

    api_key = (
        settings.embedding_api_key.get_secret_value()
        if settings.embedding_api_key
        else None
    )

    # Shared and lazy: the local model is 2.3 GB, and the query-fallback path
    # and a local bulk path must not each load their own copy.
    _local_cache: dict[str, EmbeddingPort] = {}

    def local_embedding() -> EmbeddingPort:
        if "e" not in _local_cache:
            _local_cache["e"] = LocalBGEEmbedding(
                settings.embedding_model, settings.embedding_dim
            )
        return _local_cache["e"]

    closeables: list[Any] = []
    embed_mode = _resolve_mode(
        settings.embedding_provider, base_url, what="embedding"
    )

    if embed_mode == "local_bge":
        query_embedding = bulk_embedding = local_embedding()
        embed_summary = f"embedding local ({settings.embedding_model})"
    else:
        assert base_url is not None  # _resolve_mode guarantees it
        remote = RemoteEmbeddingClient(
            base_url,
            settings.embedding_model,
            settings.embedding_dim,
            timeout=settings.embedding_timeout,
            api_key=api_key,
        )
        closeables.append(remote)
        # Bulk never falls back, in either remote mode. A corpus run that
        # silently moved to the laptop would finish next week.
        bulk_embedding = remote
        if embed_mode == "auto":
            query_embedding = QueryEmbeddingWithFallback(
                remote, local_embedding, base_url=base_url
            )
            embed_summary = f"embedding {base_url}, queries fall back to local"
        else:
            query_embedding = remote
            embed_summary = f"embedding {base_url} (no fallback)"

    rerank_mode = _resolve_mode(
        settings.reranker_provider, base_url, what="reranker"
    )

    if rerank_mode == "none":
        reranker: RerankerPort = NoopReranker()
        rerank_summary = "reranking disabled"
    elif rerank_mode == "local_bge":
        reranker = LocalBGEReranker(settings.reranker_model)
        rerank_summary = f"reranking local ({settings.reranker_model})"
    else:
        assert base_url is not None
        reranker = RemoteReranker(
            base_url,
            settings.reranker_model,
            timeout=settings.reranker_timeout,
            api_key=api_key,
            # `remote_api` insists on that host; `auto` merely prefers it, so a
            # server too old to offer /rerank degrades rather than erroring.
            require_capability=rerank_mode == "remote_api",
        )
        closeables.append(reranker)
        # Both remote modes behave the same at call time: an outage means the
        # search returns unreranked and flagged. They differ only in what
        # happens with no URL configured, which `_resolve_mode` has settled.
        rerank_summary = f"reranking {base_url}, skipped if unreachable"

    summary = f"{embed_summary}; {rerank_summary}"
    logger.info("inference_configured", summary=summary)
    return InferenceBackends(
        query_embedding=query_embedding,
        bulk_embedding=bulk_embedding,
        reranker=reranker,
        summary=summary,
        closeables=closeables,
    )


def _resolve_mode(mode: str, base_url: str | None, *, what: str) -> str:
    """Settle a configured mode against whether a host address actually exists.

    `remote_api` without an address is a contradiction the operator has to fix.
    `auto` without one is not — it means "offload if you can", and there is
    nothing to offload to, so it resolves to local and says so.
    """
    if mode in {"local_bge", "none"}:
        if base_url:
            logger.debug(
                "inference_base_url_ignored",
                what=what,
                mode=mode,
                detail=f"RE_{what.upper()}_PROVIDER={mode} overrides the "
                f"configured host. The address is kept, not used.",
            )
        return mode
    if base_url:
        return mode
    if mode == "remote_api":
        raise ConfigurationError(
            f"RE_{what.upper()}_PROVIDER=remote_api but no host is configured. "
            f"Set RE_INFERENCE_BASE_URL to a `research-engine embed-server`, "
            f"e.g. http://john-super-server:9882 — or set the provider to "
            f"'auto' to run locally when no host is reachable."
        )
    logger.info(
        "inference_auto_resolved_local",
        what=what,
        detail="RE_INFERENCE_BASE_URL is unset, so 'auto' runs on this machine.",
    )
    return "local_bge"

File: crates/marginalia-io/src/routing.rs
//! Inference placement decisions: where compute runs and what an unreachable
//! host means.
//!
//! Ports the decision layer of `adapters/inference/routing.py` — [`Workload`],
//! [`resolve_mode`], [`FallbackState`], [`is_fallback_error`], and the summary
//! builders — without `torch` or any model code. `build_inference` itself stays
//! Python: it wires live embedding/reranker adapters, lazily loads the 2.3 GB
//! local models, and owns the shutdown list, which is composition, not a
//! decision.

use crate::errors::{Error, Result};

/// What a caller is doing, which is what decides the failure policy.
pub enum Workload {
    /// One short text, a person waiting. Degrading beats failing.
    Query,
    /// Corpus-wide, unattended, results are stored. Failing beats degrading,
    /// because a silent downgrade here costs days and nobody is watching.
    Bulk,
}

impl Workload {
    pub fn as_str(&self) -> &'static str {
        match self {
            Workload::Query => "query",
            Workload::Bulk => "bulk",
        }
    }
}

/// Settle a configured mode against whether a host address actually exists.
///
/// `remote_api` without an address is a contradiction the operator has to fix.
/// `auto` without one is not — it means "offload if you can", and there is
/// nothing to offload to, so it resolves to local and says so.
/// `local_bge`/`none` win over the URL's presence, so `local_bge` is an off
/// switch that works without deleting the host address.
///
/// An empty address counts as unset, mirroring the Python `if base_url:` test,
/// which is falsy for both `None` and `""`.
pub fn resolve_mode(mode: &str, base_url: Option<&str>, what: &str) -> Result<String> {
    if mode == "local_bge" || mode == "none" {
        return Ok(mode.to_owned());
    }
    if matches!(base_url, Some(base) if !base.is_empty()) {
        return Ok(mode.to_owned());
    }
    if mode == "remote_api" {
        return Err(Error::Config(format!(
            "RE_{what}_PROVIDER=remote_api but no host is configured. Set RE_INFERENCE_BASE_URL to a `research-engine embed-server`, e.g. http://john-super-server:9882 — or set the provider to 'auto' to run locally when no host is reachable.",
            // `what` is always an ASCII literal ("embedding"/"reranker"), so
            // this matches `str.upper()` exactly.
            what = what.to_ascii_uppercase()
        )));
    }
    Ok("local_bge".to_owned())
}

/// Whether the query-fallback warning has fired. The Python wrapper logs once
/// per process — a warning on every keystroke-fast query would bury the one
/// line that explains the slowdown — then builds the local model on first
/// need. The model cache itself stays in the wiring layer; this is only the
/// once-switch.
pub struct FallbackState {
    warned: bool,
}

impl FallbackState {
    pub fn new() -> Self {
        Self { warned: false }
    }

    /// True exactly once: the first call marks and reports, later calls stay
    /// silent.
    pub fn should_warn_and_mark(&mut self) -> bool {
        if self.warned {
            false
        } else {
            self.warned = true;
            true
        }
    }
}

impl Default for FallbackState {
    fn default() -> Self {
        Self::new()
    }
}

/// Only `EmbeddingUnavailable` falls back to local. A `ModelMismatch` is a
/// misconfiguration, not an outage — falling back here would hide the one
/// failure the handshake exists to catch, and hide it permanently, since a
/// mismatch never heals on its own.
pub fn is_fallback_error(err: &Error) -> bool {
    matches!(err, Error::EmbeddingUnavailable(_))
}

/// `embedding local ({model})`: the query and bulk paths share one local
/// model, and the summary names it.
pub fn embedding_local_summary(model: &str) -> String {
    format!("embedding local ({model})")
}

/// `embedding {base}, queries fall back to local`: `auto` with a host. Bulk
/// never falls back, in either remote mode — a corpus run that silently moved
/// to the laptop would finish next week.
pub fn embedding_fallback_summary(base_url: &str) -> String {
    format!("embedding {base_url}, queries fall back to local")
}

/// `embedding {base} (no fallback)`: `remote_api` with a host.
pub fn embedding_no_fallback_summary(base_url: &str) -> String {
    format!("embedding {base_url} (no fallback)")
}

/// `reranking disabled`: `RE_RERANKER_PROVIDER=none`.
pub fn rerank_disabled_summary() -> String {
    "reranking disabled".to_owned()
}

/// `reranking local ({model})`.
pub fn rerank_local_summary(model: &str) -> String {
    format!("reranking local ({model})")
}

/// `reranking {base}, skipped if unreachable`: both remote modes behave the
/// same at call time — an outage means the search returns unreranked and
/// flagged. They differ only in what happens with no URL configured, which
/// `resolve_mode` has already settled.
pub fn rerank_remote_summary(base_url: &str) -> String {
    format!("reranking {base_url}, skipped if unreachable")
}

#[cfg(test)]
mod tests {
    use super::*;

    const HOST: &str = "http://gpu-host:9882";

    #[test]
    fn routing_workload_names_match_strenum_values() {
        assert_eq!(Workload::Query.as_str(), "query");
        assert_eq!(Workload::Bulk.as_str(), "bulk");
    }

    #[test]
    fn routing_local_mode_ignores_a_configured_host() {
        assert_eq!(
            resolve_mode("local_bge", Some(HOST), "embedding").expect("local"),
            "local_bge"
        );
        assert_eq!(
            resolve_mode("none", Some(HOST), "reranker").expect("none"),
            "none"
        );
    }

    #[test]
    fn routing_host_keeps_the_configured_mode() {
        assert_eq!(
            resolve_mode("auto", Some(HOST), "embedding").expect("auto"),
            "auto"
        );
        assert_eq!(
            resolve_mode("remote_api", Some(HOST), "embedding").expect("remote"),
            "remote_api"
        );
    }

    #[test]
    fn routing_remote_without_a_host_is_a_configuration_error() {
        for (base_url, what, provider) in [
            (None, "embedding", "EMBEDDING"),
            (None, "reranker", "RERANKER"),
            (Some(""), "embedding", "EMBEDDING"),
        ] {
            let err = resolve_mode("remote_api", base_url, what).expect_err("must fail");
            assert_eq!(
                err.to_string(),
                format!(
                    "RE_{provider}_PROVIDER=remote_api but no host is configured. Set RE_INFERENCE_BASE_URL to a `research-engine embed-server`, e.g. http://john-super-server:9882 — or set the provider to 'auto' to run locally when no host is reachable."
                ),
                "for {what:?}"
            );
        }
    }

    #[test]
    fn routing_auto_without_a_host_runs_local() {
        assert_eq!(
            resolve_mode("auto", None, "embedding").expect("auto"),
            "local_bge"
        );
        assert_eq!(
            resolve_mode("auto", Some(""), "reranker").expect("auto"),
            "local_bge"
        );
    }

    #[test]
    fn routing_fallback_warns_once_then_stays_silent() {
        let mut state = FallbackState::new();
        assert!(state.should_warn_and_mark());
        assert!(!state.should_warn_and_mark());
        assert!(!state.should_warn_and_mark());
        let mut state = FallbackState::default();
        assert!(state.should_warn_and_mark());
    }

    #[test]
    fn routing_only_unreachable_falls_back_never_mismatch() {
        assert!(is_fallback_error(&Error::EmbeddingUnavailable(
            "host is asleep".to_owned()
        )));
        assert!(!is_fallback_error(&Error::ModelMismatch {
            expected: "bge-m3".to_owned(),
            actual: "e5-large".to_owned(),
            kind: "embedding".to_owned(),
        }));
        assert!(!is_fallback_error(&Error::RerankUnavailable(
            "down".to_owned()
        )));
        assert!(!is_fallback_error(&Error::Config("bad".to_owned())));
    }

    #[test]
    fn routing_summaries_are_byte_exact() {
        assert_eq!(
            embedding_local_summary("BAAI/bge-m3"),
            "embedding local (BAAI/bge-m3)"
        );
        assert_eq!(
            embedding_fallback_summary(HOST),
            "embedding http://gpu-host:9882, queries fall back to local"
        );
        assert_eq!(
            embedding_no_fallback_summary(HOST),
            "embedding http://gpu-host:9882 (no fallback)"
        );
        assert_eq!(rerank_disabled_summary(), "reranking disabled");
        assert_eq!(
            rerank_local_summary("BAAI/bge-reranker-v2-m3"),
            "reranking local (BAAI/bge-reranker-v2-m3)"
        );
        assert_eq!(
            rerank_remote_summary(HOST),
            "reranking http://gpu-host:9882, skipped if unreachable"
        );
    }
}
